lib/shared/addon/components/new-edit-project/component.js

```diff
@@ -0,0 +1,148 @@
+import NewOrEdit from 'ui/mixins/new-or-edit';
+import layout from './template';
+
+export default Component.extend(NewOrEdit, {
+  layout,
+  intl: service(),
+  router: service(),
+  authzStore: service('authz-store'),
+  model: null,
+
+  primaryResource: alias('model.project'),
+  memberArray: alias('model.projectRoleTemplateBindings'),
+
+  actions: {
+    cancel() {
+      this.goBack();
+    },
+    addMember(kind) {
+      this.get('memberArray').pushObject({
+        subjectKind: kind,
+        subjectName: '',
+        projectRoleTemplateId: '',
+        projectId: '',
+      });
+    },
+    removeMember(obj) {
+      this.get('memberArray').removeObject(obj);
+    },
+  },
+
+  goBack: function () {
+    this.get('router').transitionTo('/projects');
+  },
+
+  doesNameExist() {
+    const project = this.get('primaryResource');
+    const currentProjects = this.get('model.projects');
+
+    if (currentProjects.findBy('name', project.get('name'))) {
+      return true;
+    }
+
+    return false;
+  },
+
+  doseMemberNameInvalid() {
+    const members = this.get('memberArray');
+    return members.any(r => r.subjectName.length === 0);
+  },
+
+  doseMemberRoleInvalid() {
+    const members = this.get('memberArray');
+    return members.any(r => r.projectRoleTemplateId.length === 0);
+  },
+
+  validate: function () {
+    var errors = this.get('errors', errors) || [];
+
+    if ((this.get('model.project.name') || '').trim().length === 0) {
+      errors.push(this.get('intl').findTranslationByKey('projectsPage.new.errors.nameReq'));
+    }
+
+    if (!this.get('editing') && this.doesNameExist()) {
+      errors.push(this.get('intl').findTranslationByKey('projectsPage.new.errors.nameInExists'));
+    }
+
+    if (this.doseMemberNameInvalid()) {
+      errors.push(this.get('intl').findTranslationByKey('projectsPage.new.errors.memberNameReq'));
+    }
+
+    if (this.doseMemberRoleInvalid()) {
+      errors.push(this.get('intl').findTranslationByKey('projectsPage.new.errors.memberRoleReq'));
+    }
+
+    if (errors.length) {
+      this.set('errors', errors.uniq());
+      return false;
+    }
+    else {
+      this.set('errors', null);
+    }
+
+    return true;
+  },
+
+  doSave() {
+    if (this.get('editing')) {
+      return this.setMembers(this.get('primaryResource'));
+    } else {
+      return this._super.apply(this, arguments).then((project) => {
+        return this.setMembers(project);
+      });
+    }
+  },
+
+  doneSaving() {
+    this.goBack();
+  },
+
+  setMembers(project) {
+    const projectId = project.id;
+    const members = this.get('memberArray');
+    const promises = [];
+    // TODO - wait for setMembers API suppot.
+    return this.get('authzStore').findAll('projectRoleTemplateBinding', { forceReload: true })
+      .then(bindings => {
+        const currentBindings = bindings.filter(b => b.projectId === projectId);
+
+        currentBindings.forEach(currentMember => {
```

Comment:
Do the adds before the deletes (though this won't matter if we add a setMembers() action).

Doing delete and then add could result in locking yourself out by deleting the binding that's allowing you add bindings and then not being able to add the rest.  (i.e. if you had "Logan has role Admin", removed that, and added "[Group Logan is in] has role Admin")